src/main/java/com/google/api/gax/retrying/AbstractRetryHandler.java

```diff
@@ -0,0 +1,137 @@
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.retrying;
+
+import com.google.api.gax.core.NanoClock;
+import com.google.api.gax.core.RetrySettings;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import org.joda.time.Duration;
+
+/**
+ * Basic implementation of the retry {@link RetryHandler} interface. Responsible for defining and
+ * checking attempt's basic properties (execution time and count limits).
+ *
+ * This class is thread-safe, and all inheriting classes are required to be thread-safe.
+ *
+ * @param <ResponseT> response type
+ */
+public abstract class AbstractRetryHandler<ResponseT> implements RetryHandler<ResponseT> {
+
+  private final NanoClock clock;
+
+  protected AbstractRetryHandler(NanoClock clock) {
+    this.clock = clock;
+  }
+
+  /**
+   * Ensures that the retry logic hasn't exceeded neither maximum number of retries nor the total
+   * execution timeout.
+   *
+   * @param e exception thrown by the previous attempt
+   * @param nextAttemptSettings attempt settings, which will be used for the next attempt, if
+   * accepted
+   * @return {@code true} if none of the retry limits are exceeded
+   */
+  @Override
+  public boolean accept(Throwable e, RetryAttemptSettings nextAttemptSettings) {
+    RetrySettings globalSettings = nextAttemptSettings.getGlobalSettings();
+    long randRetryDelayMillis = nextAttemptSettings.getRandomizedRetryDelay().getMillis();
+    long totalTimeSpentNanos =
+        clock.nanoTime()
+            - nextAttemptSettings.getFirstAttemptStartTime()
+            + TimeUnit.NANOSECONDS.convert(randRetryDelayMillis, TimeUnit.MILLISECONDS);
+
+    long totalTimeoutMillis = globalSettings.getTotalTimeout().getMillis();
+    long totalTimeoutNanos =
+        TimeUnit.NANOSECONDS.convert(totalTimeoutMillis, TimeUnit.MILLISECONDS);
+
+    return totalTimeSpentNanos <= totalTimeoutNanos
+        && (globalSettings.getMaxAttempts() <= 0
+            || nextAttemptSettings.getAttemptCount() < globalSettings.getMaxAttempts());
+  }
```

Comment:
I think there are 3 logics here:
1) Logic determining if we should retry. It is based on:
    1.1) Previous values
    1.2) Current time (not directly related to the settings, should be passed as parameter)
    1.3) Previous attempt exception (not related to the settings either, should be passed as parameter). Also exception checking is not as simple as just checking for exception type or returned code, just look at the ExceptionHandler in gcj, which is 300 lines of pretty sophisticated stuff.
2) Determining what the next attempt settings should be (because of that case, when we retry immediately if grpc deadline is exceeded, that logic is conditional, and is different for rgpc and non-grpc)
3) Executing the attempt.

And here the last one is the simplest part. If I take everything except the execution from RetryHandler and move it to RetryAttemptSettings, then RetryAttemptSettings will practically become what RetryHandler is now, but with a different name. 

Having a separate class for execution only felt like an overkill (but still cannot easily use ScheduledExecutorService directly, because we need a direct (same thread) execution, and there is no version of MoreExecutors.directExecutor() for Scheduled executor).

Also, please note, that some of these design decision were dictated by the gcj integration part. Please check how this handler is used in the related gcj pull request.

Please clarify how strong you are about making this change. I would still prefer not doing it, but if you feel really strong about it, I'm fine with making the change (in that case, please suggest names for the RetryAttemtSettings, as CallAttemptSettings does not feel like a class with lots of logic in it). Note, it will require changes in gcj integration too (basically having something like gax and gcj AttemptSettings, which I believe defeats the purpose of the whole change)